designs/node-overlay.md

```diff
@@ -0,0 +1,305 @@
+
+```
+kind: NodeOverlay
+metadata:
+  name: memory-1
+spec:
+  weight: 90
+  requirements:
+  - key: karpenter.k8s.aws/instance-memory
+    operator: Gt
+    values: ["m5.large", "m5.2xlarge"]
+  capacity:
+    memory: 50Mi
+status:
+  condition:
+  - lastTransitionTime: "..."
+    message: "conflict with overlay capacity-1"
+    reason: "conflict"
+    status: "False"
+    type: Ready
+---
+kind: NodeOverlay
+metadata:
+  name: capacity-1
+spec:
+  weight: 90
+  requirements:
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ["c5.large", "c5.2xlarge", "m5.2xlarge"]
+  capacity:
+    memory: 100Mi
+status:
+  condition:
+  - lastTransitionTime: "..."
+    message: ""
+    reason: ...
+    status: "True"
+    type: Ready
+```
+
+**Fail Closed Approach:**
+
+* Halts new node launches until misconfigurations are resolved
+* Can be implemented at two levels:
+
+    1. Global: Stops all provisioning across the cluster
+    2. NodePool-specific: Only affects misconfigured NodePools
+
+* Similar to Karpenter's existing behavior with .spec.limits and InsufficientCapacityErrors
+* Provides strong guarantees and explicit failure signals
+
+The fail closed approach, while providing strong guarantees and explicit failure signals, has significant operational drawbacks. By halting node launches until misconfigurations are resolved (either globally or per NodePool), it could unnecessarily impact cluster scalability and workload availability. This aggressive stance on configuration errors could turn minor overlay misconfigurations into broader service disruptions, particularly in scenarios where the misconfiguration affects only a subset of resources that aren't critical to immediate scaling needs. Since pods requiring specific overlay-defined resources would fail to schedule anyway (as they do today), the additional protection of failing closed provides minimal benefit while increasing operational risk.
+
+## Observability 
+
+Observability for Node Overlay will center around providing users with practical tools to understand the impact of their configurations. We will develop a diagnostic tool that allows users to preview how their Node Overlay configurations affect instance pricing and capacity allocations before deployment. This tool will help operators visualize and validate changes to instance selection patterns, price modifications, and capacity adjustments without requiring extensive logging or event generation. While we maintain basic metrics for overlay application status, this proactive tooling approach provides more actionable insights than high-volume logs or events, which could create operational overhead and scalability concerns in large clusters.
+
+```
+kind: NodeOverlay
+metadata:
+  name: price-modification
+spec:
+  requirements:
+  - key: karpenter.sh/capacity-type
+    operator: In
+    values: ["on-demand"]
+  priceAdjustment: 0.101 
```

Comment:
I'd avoid floats in APIs. They're notorious for math issues: https://stackoverflow.com/questions/36043597/why-is-1-001-0-001-not-equal-to-1-002

For this reason, you'd be hard pressed to see them in any k8s API. 

https://github.com/kubernetes-sigs/controller-tools/issues/245